### PR TITLE
Use globally consistent overmap positions for forest placement

### PR DESCRIPTION
**Pull request by ralreegorganon**
#### Summary

SUMMARY: Infrastructure "Use globally consistent overmap positions for forest placement"

#### Purpose of change

The current mechanism for placing forests on the overmap makes for extremely obvious overmap boundaries. Here we have a 2x2 overmap layout:

o_10_old_and_busted

This changes that so at least the forests are a bit more seamless.

#### Describe the solution

* Update forest placement to be noise based, indexed by the global overmap terrain coordinates.
* Update swamp placement to be partially noise based and consider river floodplains.
* Update region settings to remove the explicit settings for forests and swamps and replace them with parameters that control how the noise is interpreted.
* Update region settings documentation.

The result looks like this:

o_10_after

#### Describe alternatives you've considered

The existing forest placement could possibly be retained and patched over-ish by inspecting adjacent overmap boundaries and trying to continue the forest from any location that has an adjacent forest.

#### Additional context

Adjusting the thresholds in the region settings lets us more smoothly transition between regions (when that becomes a reality) and also lets us scale the amount of forest up and down.

**minimal forest**
o_10_minimal

**dense forest**
o_10_more_trees

**something in between**
o_10_inbetween

I tried to pick numbers and noise operations such that the approximate density and distribution matches the current game setup, but it won't be exactly the same.

One issue that remains with the swamps is that they can still show obvious overmap boundaries because they're not purely noise based--I let the proximity to rivers influence them and river placement is not currently deterministic. For the most part, this isn't as noticeable as it was before, but if you crank up the numbers so that lots of river-influenced swamps are placed, it can become quite obvious like the following:

o_10_swampthing

I still think this is an improvement overall though.

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
We need your program for creating big maps#

**Comment by Inglonias** [comment]
Is very pretty. I like it

**Comment by nexusmrsep** [comment]
Two ideas emerged in my mind when I look at these maps, for future consideration:

[ 1 ] Some cities, even in dense forests, would often have some areas around them deforested. Not totally tree-free zone, rather swathes of land where forest was removed to give land for the city's growth / communal purposes. Could be some noise-based 'deforester' centered at the city, or a nullification process that would null forests on the outskirts at random, by adding empty fields instead of building on tiles previously populated with forest. Often those deforested areas would be rural/farming plots/fields/areas.

[ 2 ] The barren fields of the minimum-trees version of the map, literally asks the question: Where are the rural/farming plots/fields/areas? Their absence never visually popped as forest were plenty, but without them it becomes obvious. Perhaps a mapgen method can be created to represent them.

Since one picture can say thousand words, I'll drop a picture form Google Maps of New England area, with said features: land taken for farming purposes (right side), some deforestation seen where city meets the forest (left side)

image

And I love the idea of seamless maps. Great job!

**Comment by codemime** [comment]
This presumably resolves #24697 (through implementing https://github.com/CleverRaven/Cataclysm-DDA/issues/24697#issuecomment-411525629).

Good job! One observation, however: the original algorithm was generating a bit more sophisticated forest edges (more curvy, with glades, ravines, and such). The simplexnoise-based algorithm produces mostly solid patches with smoother edges. Maybe it needs more octaves or finer tuning?

**Comment by ralreegorganon** [comment]
> We need your program for creating pictures of such a big maps!

It's [right here](https://github.com/ralreegorganon/cddamap). It's a command line utility, written in Go, some assembly required. You probably have no use for most of the functionality in it (serving up map data as GeoJson, rendering image tiles for use in web apps, backed by a PostgreSQL database with PostGIS, etc in support of https://cddamapalt.ralreegorganon.com and https://cddamap.ralreegorganon.com/ ) but you can use the image generator part of it just by following the instructions [here](https://github.com/ralreegorganon/cddamap#generating-an-image-from-a-world-save).

> Two ideas emerged in my mind when I look at these maps, for future consideration:
> [ 1 ] Some cities, even in dense forests, would often have some areas around them deforested. 

Yeah, I'd like to move that sort of behavior to the city building logic--it does make sense for it to happen, but it just makes more sense for me to be associated with the city rather than the forest placement.

> [ 2 ] The barren fields of the minimum-trees version of the map, literally asks the question: Where are the rural/farming plots/fields/areas? 

For sure--I think this is part of why the "desert test region" is still very much a test, and also something that we'll be able to improve with the ongoing work to use land use classification codes to drive some of the overmap generation. Also--there are plenty of overmap locations and specials that we have a hard time fitting in when the forests go away.

> This presumably resolves #24697 (through implementing #24697 (comment)).

I debated whether or not to mark it as resolved by this--this does address the forest/swamp part of it, but most other features are still overmap-bound and cause an obvious demarcation. I'm sorta burning the candle at both ends here, since the other suggestion in that comment about real world data is also [one of my side projects](https://cddamapalt.ralreegorganon.com).

> One observation, however: the original algorithm was generating a bit more sophisticated forest edges (more curvy, with glades, ravines, and such). The simplexnoise-based algorithm produces mostly solid patches with smoother edges. Maybe it needs more octaves or finer tuning?

Yeah, I could fiddle with the math for a long time trying to get closer to the older behavior (one of my other side-projects is a [world-generator based on some of the same techniques](https://ojo.ralreegorganon.com/) , but I figured it was probably worth getting the infrastructure in place now with a reasonable approximation and I or someone else could iterate on the maths subsequently. Since you mentioned it, however, I adjusted the math some to be a bit less smooth, so now it's generating something like this:

o_10_moar_updated_noise

I think it looks better, or at least closer to the old version. There are different issues with the new maths, e.g. when reducing the threshold significantly to try and create very dense forests there will still be clearings that wouldn't have been there with the previous approach, but tweaking the maths is always going to be an ongoing thing.

File: src/regional_settings.h [hunk 1]
```diff
@@ -225,4 +229,4 @@ void reset_region_settings();
 void load_region_overlay( JsonObject &jo );
 void apply_region_overlay( JsonObject &jo, regional_settings &region );
 
-#endif
+#endif
```
Comment (ZhilkinSerg) [inline src/regional_settings.h]:
```suggestion
#endif

```

File: src/regional_settings.cpp [hunk 2]
```diff
@@ -908,4 +911,4 @@ void building_bin::finalize()
     }
 
     finalized = true;
-}
+}
```
Comment (ZhilkinSerg) [inline src/regional_settings.cpp]:
```suggestion
}

```

**Pull request merged**